### PR TITLE
Minimize missed rule group evaluations

### PR DESCRIPTION
**Pull request by rajagopalanand**
**What this PR does**:

Currently once a Ruler instance loads a rule group, it evaluates it continuously. If the instance evaluating the rule group becomes unavailable, there is a high chance for missed evaluations before the instance becomes available again or another instance loads and evaluates the rule group. Ruler instances can become unavailable for variety of reasons including bad underlying nodes and OOM kills.  This issue can be exacerbated if the Ruler instance appears as healthy within the cluster ring but is actually in an unhealthy state.

This PR addresses the problem by introducing a check to ensure that the primary Ruler is alive and in a running state. Here’s how it works:

1. Liveness Check: Non-primary Rulers will perform a liveness check on the primary Ruler for each rule group when syncing rule groups from external storage.
2. Fallback Mechanism: If the primary Ruler is unresponsive or not in a running state, the non-primary Ruler will assume ownership of the rule group and take over its evaluation.
3. Relinquish Ownership: If the primary Ruler is alive and running, and if non-primary ruler has ownership of the rule group, then it relinquishes ownership of that rule group by not taking ownership and unloading it from Prometheus rule manager

With this change, the maximum duration of missed evaluations will be limited to the sync interval of the rule groups, reducing the impact of primary Ruler unavailability.

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: docs/configuration/config-file-reference.md [hunk 1]
```diff
@@ -4371,6 +4375,18 @@ ring:
 # Disable the rule_group label on exported metrics
 # CLI flag: -ruler.disable-rule-group-label
 [disable_rule_group_label: <boolean> | default = false]
+
+# Enable high availability
+# CLI flag: -ruler.enable-ha
+[enable_ha: <boolean> | default = false]
```
Comment (yeya24) [inline docs/configuration/config-file-reference.md:4381]: `enable_ha` feels like a bad name to me. Can we just replace this with checking replication factor?
If we can't, can we be specific about what HA we want to enable? For example, `enable_ha_evaluation`

File: docs/configuration/config-file-reference.md [hunk 2]
```diff
@@ -4371,6 +4375,18 @@ ring:
...
+
+# Timeout for fanout calls to other rulers
+# CLI flag: -ruler.list-rules-fanout-timeout
+[list_rules_fanout_timeout: <duration> | default = 2m]
+
+# Timeout for liveness checks performed during rule sync
+# CLI flag: -ruler.liveness-check-timeout
+[liveness_check_timeout: <duration> | default = 1s]
```
Comment (yeya24) [inline docs/configuration/config-file-reference.md:4389]: Is 1s a good default here?

File: pkg/ruler/client_pool_test.go [hunk 3]
```diff
@@ -5,6 +5,8 @@ import (
 	"net"
 	"testing"
 
+	"github.com/cortexproject/cortex/pkg/util/services"
```
Comment (yeya24) [inline pkg/ruler/client_pool_test.go:8]: Please group import

File: pkg/ruler/ruler.go [hunk 4]
```diff
@@ -273,6 +281,13 @@ type MultiTenantManager interface {
 //	|                              +-----------------+              |
 //	|                                                               |
 //	+---------------------------------------------------------------+
+
+type ruler interface {
```
Comment (yeya24) [inline pkg/ruler/ruler.go:285]: Is this nencessary

File: pkg/ruler/ruler.go [hunk 5]
```diff
@@ -488,11 +517,11 @@ func tokenForGroup(g *rulespb.RuleGroupDesc) uint32 {
 	return ringHasher.Sum32()
 }
 
-func instanceOwnsRuleGroup(r ring.ReadRing, g *rulespb.RuleGroupDesc, disabledRuleGroups validation.DisabledRuleGroups, instanceAddr string, forBackup bool) (bool, error) {
+func instanceOwnsRuleGroup(r ruler, rr ring.ReadRing, g *rulespb.RuleGroupDesc, disabledRuleGroups validation.DisabledRuleGroups, instanceAddr string, forBackup bool) (bool, error) {
```
Comment (yeya24) [inline pkg/ruler/ruler.go:520]: Should we just use `*Ruler` as the method receiver here?

File: pkg/ruler/ruler.go [hunk 6]
```diff
@@ -507,8 +536,24 @@ func instanceOwnsRuleGroup(r ring.ReadRing, g *rulespb.RuleGroupDesc, disabledRu
...
+				}
+				if ruler.Addr == instanceAddr && nonPrimaryInstanceOwnsRuleGroup(r, g, rlrs.GetAddresses()[:i], instanceAddr) {
+					level.Info(r.Logger()).Log("msg", "non-primary ruler taking ownership", "user", g.User, "group", g.Name, "namespace", g.Namespace, "token", hash, "ruler", instanceAddr)
+					ownsRuleGroup = true
+					break
+				}
+			}
+		} else {
+			// Even if the replication factor is set to a number bigger than 1, only the first ruler evaluates the rule group
+			ownsRuleGroup = rlrs.Instances[0].Addr == instanceAddr
```
Comment (yeya24) [inline pkg/ruler/ruler.go:555]: Is this else statement necessary? I see we still keep L539 above

File: pkg/ruler/ruler.go [hunk 7]
```diff
@@ -507,8 +536,24 @@ func instanceOwnsRuleGroup(r ring.ReadRing, g *rulespb.RuleGroupDesc, disabledRu
 			}
 		}
 	} else {
-		// Even if the replication factor is set to a number bigger than 1, only the first ruler evaluates the rule group
 		ownsRuleGroup = rlrs.Instances[0].Addr == instanceAddr
+		if r.Config().EnableHA {
+			for i, ruler := range rlrs.Instances {
+				if ruler.Addr == instanceAddr && i == 0 {
+					level.Debug(r.Logger()).Log("msg", "primary taking ownership", "user", g.User, "group", g.Name, "namespace", g.Namespace, "token", hash, "ruler", instanceAddr)
```
Comment (yeya24) [inline pkg/ruler/ruler.go:543]: I wonder if it really helps to log `token` here. Does it help debug HA issues?

File: pkg/ruler/ruler.go [hunk 8]
```diff
@@ -518,6 +563,68 @@ func instanceOwnsRuleGroup(r ring.ReadRing, g *rulespb.RuleGroupDesc, disabledRu
...
+	if r.lifecycler.ServiceContext().Err() != nil || r.subservices.IsStopped() {
+		return nil, errors.New("ruler's context is canceled and might be stopping soon")
+	}
+	if !r.subservices.IsHealthy() {
+		return nil, errors.New("not all subservices are in healthy state")
+	}
+	return &LivenessCheckResponse{State: int32(r.State())}, nil
+}
+
+func nonPrimaryInstanceOwnsRuleGroup(r ruler, g *rulespb.RuleGroupDesc, replicas []string, selfAddress string) bool {
```
Comment (yeya24) [inline pkg/ruler/ruler.go:577]: It would be great to have some comments on this function for better readability

File: pkg/ruler/ruler.go [hunk 9]
```diff
@@ -1164,6 +1285,8 @@ func (r *Ruler) getShardedRules(ctx context.Context, userID string, rulesRequest
 		}
 	}
 	// Concurrently fetch rules from all rulers.
+	ctx, cancel := context.WithTimeout(ctx, r.cfg.ListRulesFanoutTimeout)
```
Comment (yeya24) [inline pkg/ruler/ruler.go:1288]: This will be a behavior change for all Cortex users. Can we just use the API request's own timeout?

File: pkg/ruler/ruler.go [hunk 10]
```diff
@@ -518,6 +563,68 @@ func instanceOwnsRuleGroup(r ring.ReadRing, g *rulespb.RuleGroupDesc, disabledRu
 	return ownsRuleGroup, nil
 }
 
+func (r *Ruler) LivenessCheck(_ context.Context, request *LivenessCheckRequest) (*LivenessCheckResponse, error) {
+	level.Debug(r.logger).Log("msg", "liveness check request", "request received from", request.RulerAddress)
```
Comment (yeya24) [inline pkg/ruler/ruler.go:567]: Seems `RulerAddress` is only used here for debug logging purpose. Can we remove that?

File: pkg/ruler/ruler.go [hunk 11]
```diff
@@ -518,6 +563,68 @@ func instanceOwnsRuleGroup(r ring.ReadRing, g *rulespb.RuleGroupDesc, disabledRu
...
+	if r.lifecycler.ServiceContext().Err() != nil || r.subservices.IsStopped() {
+		return nil, errors.New("ruler's context is canceled and might be stopping soon")
```
Comment (yeya24) [inline pkg/ruler/ruler.go:569]: I am unsure about this tbh. When `r.lifecycler.ServiceContext().Err() != nil || r.subservices.IsStopped()` condition is met, is the gRPC server still working correctly to handle liveness check?

Comment (rajagopalanand) [reply pkg/ruler/ruler.go:569]: Depending upon how Ruler is configured, `r.lifecycler.ServiceContext()` can be canceled but gRPC server can be active for a long time. When `r.lifecycler.ServiceContext()` is canceled, Prometheus will stop evaluating rule groups as soon as possible but if the Ruler instance has not shutdown yet, it will respond to secondary as though it is active and this causes missed evaluations. This check is necessary.

`r.subservices.IsStopped()` is debatable. I only saw this in my testing in some cases during hard restarts (kill -9) when the gRPC server was already running but ring was not fully operational

File: pkg/ruler/ruler.go (hunk 10 above)
Comment (rajagopalanand) [reply pkg/ruler/ruler.go:567]: The whole log is a debug log line and the log is not useful if it does not have the ruler address. If someone wants to troubleshoot, having the ruler address that performed the liveness check would be useful I think.

File: pkg/ruler/ruler.go (hunk 7 above)
Comment (rajagopalanand) [reply pkg/ruler/ruler.go:543]: It helps issues with troubleshooting evaluation issues. Users can filter logs by token and know what happened with a particular rule group

File: pkg/ruler/ruler.go (hunk 5 above)
Comment (rajagopalanand) [reply pkg/ruler/ruler.go:520]: Reasons for introducing an interface is to narrow down what is available for functions to consume. Existing functions such as `filterRuleGroups` is not a method to avoid accidentally using Ruler's ring directly. If I pass in Ruler struct/object as-is, then it will expose the ring to those functions which could lead to accidental use of it. This is the reason I introduced a interface so that it could narrow down the scope of what is available to these functions. Of course, someone in the future could just add a method on the interface that exposes the ring. Having said all this, I'm open to suggestions here

File: pkg/ruler/ruler.go (hunk 4 above)
Comment (rajagopalanand) [reply pkg/ruler/ruler.go:285]: Please see https://github.com/cortexproject/cortex/pull/6129#discussion_r1704241686

File: pkg/ruler/ruler.go (hunk 6 above)
Comment (rajagopalanand) [reply pkg/ruler/ruler.go:555]: Good catch! The `else` is necessary but L539 must be removed

File: docs/configuration/config-file-reference.md (hunk 1 above)
Comment (rajagopalanand) [reply docs/configuration/config-file-reference.md:4381]: API HA checks only for replication factor. My thought was to use a separate flag in case if someone wants to use one or the other and not both. I can change this to `enable_ha_evaluation`

File: pkg/ruler/ruler.go (hunk 9 above)
Comment (rajagopalanand) [reply pkg/ruler/ruler.go:1288]: If the client sets the timeout, then this timeout/deadline will also be the same and it will fail the overall API request for the client. The desired behavior is not to fail the whole API request but instead be fault tolerant if a subset of Rulers do not respond in a certain amount of time. Ideally,  `r.cfg.ListRulesFanoutTimeout` should be less than the API request's timeout. If we do not want to change existing behavior, we can either set the default timeout to be a much bigger number (max int) or wrap this around the `enable_ha_evaluation` flag or perhaps there is a better option. Open to recommendations

File: docs/configuration/config-file-reference.md (hunk 2 above)
Comment (rajagopalanand) [reply docs/configuration/config-file-reference.md:4389]: This is a simple/fast check. I thought `1s` might be big enough for default. If RF=2, this should check only one Ruler. If RF=3, it will check 2 Rulers. Do you think it needs to be higher/lower?

File: pkg/ruler/ruler.go [hunk 12]
```diff
@@ -518,6 +562,70 @@ func instanceOwnsRuleGroup(r ring.ReadRing, g *rulespb.RuleGroupDesc, disabledRu
...
+			level.Debug(r.Logger()).Log("msg", "liveness check failed", "addr", addr, "for", g.Name, "err", err.Error(), "token", rgToken)
+			return nil
+		}
+		level.Debug(r.Logger()).Log("msg", "liveness check succeeded ", "addr", addr, "for", g.Name, "token", rgToken, "ruler state", services.State(resp.GetState()))
+		responseChan <- resp
+		return nil
+	})
+
+	close(errorChan)
+	close(responseChan)
```
Comment (alanprot) [inline pkg/ruler/ruler.go:615]: if one ruler is taking time to reply the liveness, wouldn't it make this whole call to hang?

File: pkg/ruler/ruler.go (hunk 10 above)
Comment (alanprot) [reply pkg/ruler/ruler.go:566]: humm we cannot log it in the client side? wouldn't this have the same effect?

File: docs/configuration/config-file-reference.md (hunk 2 above)
Comment (alanprot) [reply docs/configuration/config-file-reference.md:4399]: Do we need this config at all? We cannot just put a sensible timeout?

File: pkg/ruler/ruler.go (hunk 7 above)
Comment (yeya24) [reply pkg/ruler/ruler.go:542]: > Users can filter logs by token and know what happened with a particular rule group

Are we going to document this clearly? Otherwise I don't think it is something easy to debug as a user. Tbh I don't how I map a particular rule group to a hash by just looking at the log.
Isn't `group` and `namespace` already enough to locate a particular rule group?

File: pkg/ruler/ruler.go (hunk 6 above)
Comment (alanprot) [reply pkg/ruler/ruler.go:554]: does this needs to be inside the for??? can we on foreach the instances if `EnableHAEvaluation` is enabled and if not, we do `rlrs.Instances[0].Addr == instanceAddr` check?

File: pkg/ruler/ruler.go (hunk 12 above)
Comment (alanprot) [reply pkg/ruler/ruler.go:615]: discussed offline... i just think the timeout could be hardcoded and not a config.

File: pkg/ruler/ruler.go (hunk 5 above)
Comment (rajagopalanand) [reply pkg/ruler/ruler.go]: I changed `filterRuleGroups` and other functions to methods on the Ruler and removed the interface

File: pkg/ruler/ruler.go (hunk 7 above)
Comment (rajagopalanand) [reply pkg/ruler/ruler.go]: Removed

File: pkg/ruler/ruler.go (hunk 10 above)
Comment (rajagopalanand) [reply pkg/ruler/ruler.go]: Removed

File: pkg/ruler/ruler.go (hunk 12 above)
Comment (rajagopalanand) [reply pkg/ruler/ruler.go]: Set the timeout to 100ms and removed the config

File: pkg/ruler/ruler.go (hunk 6 above)
Comment (rajagopalanand) [reply pkg/ruler/ruler.go:550]: @alanprot It already is that way, no? if `EnableHAEvaluation` is disabled, we do `rlrs.Instances[0].Addr == instanceAddr` check outside the for loop. If `EnableHAEvaluation` is enabled, then loop thru the instances and perform the liveness check

File: pkg/ruler/ruler.go [hunk 13]
```diff
@@ -220,7 +225,12 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	f.BoolVar(&cfg.EnableQueryStats, "ruler.query-stats-enabled", false, "Report the wall time for ruler queries to complete as a per user metric and as an info level log message.")
 	f.BoolVar(&cfg.DisableRuleGroupLabel, "ruler.disable-rule-group-label", false, "Disable the rule_group label on exported metrics")
 
+	f.BoolVar(&cfg.EnableHAEvaluation, "ruler.enable-ha-evaluation", false, "Enable high availability")
+	f.DurationVar(&cfg.ListRulesFanoutTimeout, "ruler.list-rules-fanout-timeout", 2*time.Minute, "Timeout for fanout calls to other rulers")
+	f.DurationVar(&cfg.LivenessCheckTimeout, "ruler.liveness-check-timeout", 1*time.Second, "Timeout for liveness checks performed during rule sync")
+
 	cfg.RingCheckPeriod = 5 * time.Second
+	cfg.LivenessCheckTimeout = 100 * time.Millisecond
```
Comment (alanprot) [inline pkg/ruler/ruler.go:233]: can this just be a const? no need to be in the config struct i guess.

File: pkg/ruler/ruler.go [hunk 14]
```diff
@@ -220,7 +225,12 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	f.BoolVar(&cfg.EnableQueryStats, "ruler.query-stats-enabled", false, "Report the wall time for ruler queries to complete as a per user metric and as an info level log message.")
 	f.BoolVar(&cfg.DisableRuleGroupLabel, "ruler.disable-rule-group-label", false, "Disable the rule_group label on exported metrics")
 
+	f.BoolVar(&cfg.EnableHAEvaluation, "ruler.enable-ha-evaluation", false, "Enable high availability")
+	f.DurationVar(&cfg.ListRulesFanoutTimeout, "ruler.list-rules-fanout-timeout", 2*time.Minute, "Timeout for fanout calls to other rulers")
+	f.DurationVar(&cfg.LivenessCheckTimeout, "ruler.liveness-check-timeout", 1*time.Second, "Timeout for liveness checks performed during rule sync")
```
Comment (alanprot) [inline pkg/ruler/ruler.go:230]: and this should be removed as well?

File: pkg/ruler/ruler.go (hunk 9 above)
Comment (alanprot) [reply pkg/ruler/ruler.go:1281]: +1 to bens comment... I think the solution here is not to set a timeout.. instead, we should return as soon as we have a complete answer (quorum?)

File: integration/ruler_test.go [hunk 15]
```diff
@@ -1093,6 +1093,147 @@ func TestRulerDisablesRuleGroups(t *testing.T) {
...
+		var exprNode yaml.Node
+
+		ruleNode.SetString(fmt.Sprintf("rule_%d", i))
+		exprNode.SetString(strconv.Itoa(i))
+		ruleName := fmt.Sprintf("test_%d", i)
+
+		expectedNames[i] = ruleName
+
+		if num%2 == 0 {
+			alertCount++
```
Comment (rapphil) [inline integration/ruler_test.go:1121]: is this variable used anywhere?

File: integration/ruler_test.go [hunk 16]
```diff
@@ -1093,6 +1093,147 @@ func TestRulerDisablesRuleGroups(t *testing.T) {
...
+	ruler2 := e2ecortex.NewRuler("ruler-2", consul.NetworkHTTPEndpoint(), rulerFlags, "")
+	ruler3 := e2ecortex.NewRuler("ruler-3", consul.NetworkHTTPEndpoint(), rulerFlags, "")
+	rulers := e2ecortex.NewCompositeCortexService(ruler1, ruler2, ruler3)
+	require.NoError(t, s.StartAndWaitReady(ruler1, ruler2, ruler3))
+
+	// Upload rule groups to one of the rulers.
+	c, err := e2ecortex.NewClient("", "", "", ruler1.HTTPEndpoint(), "user-1")
+	require.NoError(t, err)
+	namespaceNames := []string{"test1", "test2", "test3", "test4", "test5"}
+	namespaceNameCount := make([]int, 5)
```
Comment (rapphil) [inline integration/ruler_test.go:1178]: Total nit: in case you want to add more users in the tests.

```suggestion
	namespaceNameCount := make([]int, len(namespaceNames))
```

File: integration/ruler_test.go [hunk 17]
```diff
@@ -1093,6 +1093,147 @@ func TestRulerDisablesRuleGroups(t *testing.T) {
...
+
+	countFunc(ruler1Rules)
+	require.Equal(t, numRulesGroups, ruleCount)
+	ruleCount = 0
+	countFunc(ruler3Rules)
+	require.Equal(t, numRulesGroups, ruleCount)
+
+	results, err := c.GetPrometheusRules(e2ecortex.RuleFilter{})
+	require.NoError(t, err)
+	require.Equal(t, numRulesGroups, len(results))
```
Comment (rapphil) [inline integration/ruler_test.go:1234]: I'm a bit confused on how this test is validating HA. Should we validate that the total number of validations across all rules is what is expected?

File: pkg/ruler/ruler.go [hunk 18]
```diff
@@ -518,6 +556,68 @@ func instanceOwnsRuleGroup(r ring.ReadRing, g *rulespb.RuleGroupDesc, disabledRu
...
+		rulerClient, err := r.GetClientFor(addr)
+		if err != nil {
+			errorChan <- err
+			level.Debug(r.Logger()).Log("msg", "unable to get client for ruler", "ruler addr", addr)
+			return nil
+		}
+		level.Debug(r.Logger()).Log("msg", "performing liveness check against", "addr", addr, "for", g.Name, "instance addr", selfAddress)
+
+		resp, err := rulerClient.LivenessCheck(ctx, &LivenessCheckRequest{
+			RulerAddress: selfAddress,
```
Comment (rapphil) [inline pkg/ruler/ruler.go:594]: any reason why we are passing the address of the current ruler as parameter here? I checked and this is not being used.

File: pkg/ruler/ruler.go [hunk 19]
```diff
@@ -507,8 +530,23 @@ func instanceOwnsRuleGroup(r ring.ReadRing, g *rulespb.RuleGroupDesc, disabledRu
...
+					break
+				}
+			}
+		} else {
+			// Even if the replication factor is set to a number bigger than 1, only the first ruler evaluates the rule group
+			ownsRuleGroup = rlrs.Instances[0].Addr == instanceAddr
+		}
 	}
 
 	if ownsRuleGroup && ruleGroupDisabled(g, disabledRuleGroups) {
```
Comment (rapphil) [inline pkg/ruler/ruler.go:552]: nit: we could create a function to generate the return in case of ownsRuleGroup = true.

this way we can retur directly from the code with early returns instead of using this nesting which is confusing and hard to reason about.

```golang
func (r *Ruler) instanceOwnsRuleGroup(rr ring.ReadRing, g *rulespb.RuleGroupDesc, disabledRuleGroups validation.DisabledRuleGroups, instanceAddr string, forBackup bool) (bool, error) {

	hash := tokenForGroup(g)

	rlrs, err := rr.Get(hash, RingOp, nil, nil, nil)
	if err != nil {
		return false, errors.Wrap(err, "error reading ring to verify rule group ownership")
	}

	if forBackup {
		// Only the second up to the last replica are used as backup
		for i := 1; i < len(rlrs.Instances); i++ {
			if rlrs.Instances[i].Addr == instanceAddr {
				return ownsRuleGroupOrDisable(g, disabledRuleGroups)
			}
		}
	} 
        if rlrs.Instances[0].Addr == instanceAddr {
              // regardless of ruler HA is enabled or not, in this case this ruler is the primary
               return ownsRuleGroupOrDisable(g, disabledRuleGroups)
        } 
	if r.Config().EnableHAEvaluation {
		for i, ruler := range rlrs.Instances {
			if  i == 0 {
                                 // we already checked for primary
                                  continue
			}
			if ruler.Addr == instanceAddr && r.nonPrimaryInstanceOwnsRuleGroup(g, rlrs.GetAddresses()[:i], instanceAddr) {
				level.Info(r.Logger()).Log("msg", "non-primary ruler taking ownership", "user", g.User, "group", g.Name, "namespace", g.Namespace, "ruler", instanceAddr)
				return ownsRuleGroupOrDisable(g, disabledRuleGroups)
			}
		}		
	} 
        // means for sure that the instance is not the owner
	return false, nil
}

func ownsRuleGroupOrDisable(g *rulespb.RuleGroupDesc, disabledRuleGroups validation.DisabledRuleGroups)
	if ruleGroupDisabled(g, disabledRuleGroups) {
		return false, &DisabledRuleGroupErr{Message: fmt.Sprintf("rule group %s, namespace %s, user %s is disabled", g.Name, g.Namespace, g.User)}
	}
        return true, nil
}
```

File: pkg/ruler/ruler.go (hunk 9 above)
Comment (rapphil) [reply pkg/ruler/ruler.go:1279]: I'm curious, is this line change related to minimizing missed evaluations or is it unrelated? If unrelated and you are solving another problem I would rather have this in a different pr, just because this can introduce regressions.

File: pkg/ruler/ruler.go (hunk 18 above)
Comment (rajagopalanand) [reply pkg/ruler/ruler.go:594]: It's not being used now but I was using it for troubleshooting/logging on the server side. It could still be useful for troubleshooting

**Comment by danielblando** [comment]
We are in the process of releasing 1.18
Please rebase PR and change the changelog to master

File: pkg/ruler/ruler.go (hunk 9 above)
Comment (rajagopalanand) [reply pkg/ruler/ruler.go]: > +1 to bens comment... I think the solution here is not to set a timeout.. instead, we should return as soon as we have a complete answer (quorum?)

When Ruler API HA is enabled and the rules are backed up, the backed up rules do not contain any state. In other words, it does not contain any alerts. To get a complete answer means, we have to get Rules from 100% of the Rulers. 

As an example, 3 AZs, RF = 3, ruler count per AZ = 3 (total = 9). Assume 1 ruler in AZ 2 is unhealthy. What we want is results from 8 rulers. Similarly if we have 2 bad rulers, then we need results from 7. In this scenario, we we can tolerate 2 AZs being down (RF=3) but if we bail as soon as results from 1 AZ are retrieved, then the results will contain a lot of backup rules which does not contain alerts

I moved the timeout inside the `concurrency.ForEach` so that each `Rules()` fan-out call will get its own timeout rather than the timeout being set for the whole `getShardedRules`

Let me know if this is acceptable or if there are other ideas to handle this scenario

File: integration/ruler_test.go (hunk 15 above)
Comment (rajagopalanand) [reply integration/ruler_test.go]: Removed

File: integration/ruler_test.go (hunk 17 above)
Comment (rajagopalanand) [reply integration/ruler_test.go:1121]: I updated the test to assert that the rules are evaluated by the remaining rulers

File: pkg/ruler/ruler.go (hunk 19 above)
Comment (rajagopalanand) [reply pkg/ruler/ruler.go]: done

File: pkg/ruler/ruler.go (hunk 18 above)
Comment (rapphil) [reply pkg/ruler/ruler.go:595]: If it is supposed to be used for troubleshooting we can add a debug log statement in the handler no? I think leaving in this state makes me think of YAGNI

File: docs/configuration/config-file-reference.md [hunk 20]
```diff
@@ -4340,6 +4340,10 @@ ring:
   # CLI flag: -ruler.ring.final-sleep
   [final_sleep: <duration> | default = 0s]
 
+  # Keep instance in the ring on shut down.
+  # CLI flag: -ruler.ring.keep-instance-in-the-ring-on-shutdown
+  [keep_instance_in_the_ring_on_shutdown: <boolean> | default = false]
```
Comment (alanprot) [inline docs/configuration/config-file-reference.md:4345]: Can we use the same config name as the other services? Ex:

```
# Unregister the compactor during shutdown if true.
  # CLI flag: -compactor.ring.unregister-on-shutdown
  [unregister_on_shutdown: <boolean> | default = true]
```

File: pkg/ruler/ruler.go [hunk 21]
```diff
@@ -490,34 +512,111 @@ func tokenForGroup(g *rulespb.RuleGroupDesc) uint32 {
...
+	ctx := user.InjectOrgID(context.Background(), userID)
+	ctx, cancel := context.WithTimeout(ctx, livenessCheckTimeout)
+	defer cancel()
+
+	err := concurrency.ForEach(ctx, jobs, len(jobs), func(ctx context.Context, job interface{}) error {
+		addr := job.(string)
+		rulerClient, err := r.GetClientFor(addr)
+		if err != nil {
+			errorChan <- err
+			level.Debug(r.Logger()).Log("msg", "unable to get client for ruler", "ruler addr", addr)
```
Comment (alanprot) [inline pkg/ruler/ruler.go:589]: level here should be higher than debug?

File: pkg/ruler/ruler.go [hunk 22]
```diff
@@ -593,6 +692,10 @@ func (r *Ruler) syncRules(ctx context.Context, reason string) {
 		return
 	}
 
+	if ctx.Err() != nil {
+		level.Info(r.logger).Log("msg", "context is canceled. not syncing rules")
+		return
+	}
```
Comment (alanprot) [inline pkg/ruler/ruler.go:698]: why we needed to introduce this?

File: pkg/ruler/ruler.go (hunk 18 above)
Comment (alanprot) [reply pkg/ruler/ruler.go:595]: I also thing we can remove for now and add when needed.

File: pkg/ruler/ruler.go [hunk 23]
```diff
@@ -835,12 +937,21 @@ func filterRuleGroups(userID string, ruleGroups []*rulespb.RuleGroupDesc, disabl
 // filterBackupRuleGroups returns map of rule groups that given instance backs up based on supplied ring.
 // This function only uses User, Namespace, and Name fields of individual RuleGroups.
 //
-// Reason why this function is not a method on Ruler is to make sure we don't accidentally use r.ring,
-// but only ring passed as parameter.
-func filterBackupRuleGroups(userID string, ruleGroups []*rulespb.RuleGroupDesc, disabledRuleGroups validation.DisabledRuleGroups, ring ring.ReadRing, instanceAddr string, log log.Logger, ringCheckErrors prometheus.Counter) []*rulespb.RuleGroupDesc {
+// This method must not use r.ring, but only ring passed as parameter
+func (r *Ruler) filterBackupRuleGroups(userID string, ruleGroups []*rulespb.RuleGroupDesc, owned []*rulespb.RuleGroupDesc, disabledRuleGroups validation.DisabledRuleGroups, ring ring.ReadRing, instanceAddr string, log log.Logger, ringCheckErrors prometheus.Counter) []*rulespb.RuleGroupDesc {
```
Comment (alanprot) [inline pkg/ruler/ruler.go:941]: Seems lots of parameters can be removed from this function now that is a method of the Ruler struct?

File: pkg/ruler/ruler.go (hunk 22 above)
Comment (rajagopalanand) [reply pkg/ruler/ruler.go:698]: When ruler is shutting down/pod terminating, syncs happen. This is because the manager waits for rule groups to stop finishing evaluation. I think it's better to not sync rule groups when ruler is shutting down

File: docs/configuration/config-file-reference.md (hunk 20 above)
Comment (rajagopalanand) [reply docs/configuration/config-file-reference.md:4345]: Ruler ring returns `BasicLifecyclerConfig` [here](https://github.com/cortexproject/cortex/blob/master/pkg/ruler/ruler_ring.go#L101). `BasicLifecyclerConfig` uses `KeepInstanceInTheRingOnShutdown` as shown [here](https://github.com/cortexproject/cortex/blob/master/pkg/ring/basic_lifecycler.go#L59)

Comment (alanprot) [reply docs/configuration/config-file-reference.md:4345]: Oh :/

Ok.. I guess we failed to have the same config name in the past.. =/ Very weird to have 2 configs that does the same thing but in different component.. but as we already have another one that use `keep_instance_in_the_ring_on_shutdown` i guess im fine.

File: pkg/ruler/ruler.go (hunk 22 above)
Comment (alanprot) [reply pkg/ruler/ruler.go:697]: It just weird that u check it here as this context can be cancelled any place in the code after this line as well and what u trying to prevent can happen anyway?

File: pkg/ruler/ruler.go [hunk 24]
```diff
@@ -1189,6 +1300,8 @@ func (r *Ruler) getShardedRules(ctx context.Context, userID string, rulesRequest
 			return errors.Wrapf(err, "unable to get client for ruler %s", addr)
 		}
 
+		ctx, cancel := context.WithTimeout(ctx, r.cfg.ListRulesFanoutTimeout)
```
Comment (alanprot) [inline pkg/ruler/ruler.go:1303]: I think this change is not related to this PR?

Can we remove this from here and create another one where we can discuss possible solutions for this??

**Review by alanprot — APPROVED** [review]

**Comment by alanprot** [comment]
Approved... other than changing the ruler timeout config to be under the ruler_client section, it LGTM

**Comment by rajagopalanand** [comment]
> Approved... other than changing the ruler timeout config to be under the ruler_client section, it LGTM

Pushed up a new commit. Please take a look

File: CHANGELOG.md [hunk 25]
```diff
@@ -18,6 +18,7 @@
 * [CHANGE] Querier: Remove `-querier.at-modifier-enabled` flag. #6157
 * [CHANGE] Tracing: Remove deprecated `oltp_endpoint` config entirely. #6158
 * [CHANGE] Store Gateway: Enable store gateway zone stable shuffle sharding by default. #6161
+* [FEATURE] Ruler: Minimize rule group missed evaluations via `-ruler.enable-ha` flag. #6129
```
Comment (yeya24) [inline CHANGELOG.md:21]: Need to rebase master and move this changelog entry to `master / unreleased` section

Comment (yeya24) [reply CHANGELOG.md:21]: Can we make the changelog more descriptive? As an end user, I don't understand what does it mean by `Minimize rule group missed evaluations`. This flag helps end users under which scenario? I see we added other new flags as well.

Comment (yeya24) [reply CHANGELOG.md:21]: And the flag name is `-ruler.enable-ha-evaluation`

File: integration/ruler_test.go [hunk 26]
```diff
@@ -981,6 +981,152 @@ func TestRulerDisablesRuleGroups(t *testing.T) {
 	})
 }
 
+func TestRulerHA(t *testing.T) {
```
Comment (yeya24) [inline integration/ruler_test.go:984]: Nit. `TestRulerHA` seems a bit bias. Can we call it `TestRulerHAEvaluation`?

File: docs/configuration/config-file-reference.md [hunk 27]
```diff
@@ -4142,6 +4142,10 @@ ruler_client:
   # CLI flag: -ruler.client.tls-insecure-skip-verify
   [tls_insecure_skip_verify: <boolean> | default = false]
 
+  # Timeout for downstream rulers.
```
Comment (yeya24) [inline docs/configuration/config-file-reference.md:4145]: I think we need a doc talking about Ruler HA. How to enable it with various configs/flags.
We added 3 new flags in this PR but I am confused if I need to tune/change them somehow or they should work out of the box.

**Comment by yeya24** [comment]
Note that https://github.com/cortexproject/cortex/pull/5862 needs to be merged before this PR. Without enough approvals on the proposal itself, I don't think we can merge the actual implementation.

File: CHANGELOG.md [hunk 28]
```diff
@@ -7,6 +7,7 @@
 * [ENHANCEMENT] Ruler: Add query statistics metrics when --ruler.query-stats-enabled=true. #6173
 * [ENHANCEMENT] Ingester: Add new API `/ingester/all_user_stats` which shows loaded blocks, active timeseries and ingestion rate for a specific ingester. #6178
 * [ENHANCEMENT] Distributor: Add new `cortex_reduced_resolution_histogram_samples_total` metric to to track the number of histogram samples which resolution was reduced. #6182
+* [FEATURE] Ruler: Minimize chances of missed rule group evaluations that can occur due to OOM kills, bad underlying nodes, or due to an unhealthy ruler that appears in the ring as healthy. This feature is enabled via `-ruler.enable-ha-evaluation` flag. #6129
```
Comment (yeya24) [inline CHANGELOG.md:10]: `FEATURE` needs to be put above `ENHANCEMENT`

File: pkg/ruler/ruler.go (hunk 24 above)
Comment (rajagopalanand) [reply pkg/ruler/ruler.go]: This is related as we discussed offline. Changed it to `remote_timeout` under ruler_client section

File: docs/configuration/config-file-reference.md (hunk 27 above)
Comment (rajagopalanand) [reply docs/configuration/config-file-reference.md:4145]: I agree. I will submit another PR for the doc

File: CHANGELOG.md (hunk 28 above)
Comment (rajagopalanand) [reply CHANGELOG.md]: Done

**Review by yeya24 — APPROVED** [review]
Thanks

**Review by rapphil — APPROVED** [review]

**Review by yeya24 — COMMENTED** [review]
Umm, I think we forgot to mark this as experimental. We can do it in next PR
